envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/RBACFilterFactory.kt

```diff
@@ -114,25 +114,30 @@ class RBACFilterFactory(
             loggedEndpointsPolicies.values
         )
 
-        val shadowPolicies = commonPolicies + loggedEndpointsPolicies
-        val actualPolicies = commonPolicies + allowUnlistedPolicies
+        val shadowPolicies = (commonPolicies + loggedEndpointsPolicies)
+            .map { (endpoint, policy) -> endpoint to policy.build() }.toMap()
+        val shadowRules = RBAC.newBuilder()
+            .setAction(RBAC.Action.ALLOW)
+            .putAllPolicies(shadowPolicies)
+            .build()
+        // build needs to be called before any modifications happen
 
+        val actualPolicies = (statusRoutePolicy + restrictedEndpointsPolicies.mapValues {
```

Comment:
Could you extract what is going with `restrictedEndpointsPolicies` to method?

Reply:
Or at least to a variable.